File: app/src/main/java/mariavv/fitnesspal/ui/journal/daycard/FeedAdapter.java
package mariavv.fitnesspal.ui.journal.daycard;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

class FeedAdapter extends RecyclerView.Adapter {
    private List<ItemType> dataSet = new ArrayList<>();

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        return ViewHolderFactory.create(viewGroup, viewType);
    }

    @Override
    public int getItemViewType(int position) {
        return dataSet.get(position).getItemViewType();
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int position) {
        //todo??
        dataSet.get(position).onBindViewHolder(viewHolder);
    }

    @Override
    public int getItemCount() {
        if (dataSet != null) {
            return dataSet.size();
        }
        return 0;
    }

    void updateItems(List<ItemType> entities) {
        if (entities != null) {
            dataSet = entities;
            notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/mariavv/fitnesspal/presentation/journal/daycard/FeedAdapter.kt
package mariavv.fitnesspal.presentation.journal.daycard

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import java.util.*

internal class FeedAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var dataSet: List<ItemType>? = ArrayList()

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return ViewHolderFactory.create(viewGroup, viewType)
    }

    override fun getItemViewType(position: Int): Int {
        return dataSet!![position].itemViewType
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        //todo??
        dataSet!![position].onBindViewHolder(viewHolder)
    }

    override fun getItemCount(): Int {
        return if (dataSet != null) {
            dataSet!!.size
        } else 0
    }

    fun updateItems(dataSet: List<ItemType>) {
        this.dataSet = dataSet
        notifyDataSetChanged()
    }
}
